File: DataStructures-Algorithms/5. Queues/Sliding_window_maximum_2.c
/*
Given an array and an integer K, find the maximum for each and every contiguous subarray of size k.

Time Complexity : O(n)
Space Complexity : O(k)
*/

#include<stdio.h>
#include<stdbool.h>

int peekFirst( int *deque, int *arr, int first)
{
	return arr[deque[first]];
}

int peekLast( int *deque, int *arr, int rear)
{
	return arr[deque[rear]];
}

bool isQueueEmpty(int rear)
{
	return rear == -1;
}

void removeLast(int *front, int *rear)
{
	--*rear;
	
	if(*rear == -1)
		*front = 0;
}

void removeFirst(int *front)
{
	++*front;
}

void addLast( int *deque, int *rear, int index)
{
	deque[++*rear] = index;
}

void printMax( int *arr, int n, int k)
{
	int i, front, rear, deque[k];
	
	front = 0;
	
	rear = -1;
	
	for(i=0; i<k; ++i)
	{
		while(!isQueueEmpty(rear) && arr[i] >= peekLast(deque,arr,rear))
			removeLast(&front, &rear);
		
		addLast(deque, &rear, i);
	}
	
	for( ; i<n; ++i)
	{
		printf(" %d", peekFirst(deque, arr, front));
		
		while(!isQueueEmpty(rear) && deque[front] <= i-k)
			removeFirst(&front);
		
		while(!isQueueEmpty(rear) && arr[i] >= peekLast(deque,arr,rear))
			removeLast(&front, &rear);
		
		addLast(deque, &rear, i);
	}
	
	printf(" %d", peekFirst(deque, arr, front));
}

void main()
{
	int arr[] = {8, 6, 10, 7, 5, 17, 14, 87, 15};
	
	int n = sizeof(arr)/sizeof(*arr);
	
	int k = 3;
	
	printMax(arr,n,k);
}

File: CODING/5. Queues/Sliding_window_maximum_2.java
/*
Given an array and an integer K, find the maximum for each and every contiguous subarray of size k.

Time Complexity : O(N)
Space Complexity : O(K)
*/

import java.util.*;

class GFG { 

	static void printMax(int arr[], int n, int k) {

		Deque<Integer> deque = new LinkedList<Integer>();

		int i;

		for(i = 0; i < k; ++i) {

			while(!deque.isEmpty() && arr[i] >= arr[deque.peekLast()]) {

				deque.removeLast();
			}
			deque.addLast(i);
		}

		for( ; i < n; ++i) {

			System.out.print(arr[deque.peek()] + ", ");

			while(!deque.isEmpty() && deque.peek() <= i-k) {

				deque.removeFirst();
			}

			while(!deque.isEmpty() && arr[i] >= arr[deque.peekLast()]) {

				deque.removeLast();
			}
			deque.addLast(i);
		}
		System.out.print(arr[deque.peek()] + ", ");
	}

	public static void main(String[] args) {

		int arr[] = { 12, 1, 78, 90, 57, 89, 56 };
		int k = 3;

		printMax(arr, arr.length, k);		
	}
}
